Improve u.printer/datafiable? and u.printer/navigable?

zampino opened the pull request:
At present these function only check implementations available _at def time_ i.e. when REPL is upgraded. If we extend Datafiable and Navigable protocols for some types _after_ the upgrade, such types will not pass the check even though they actually `u.p/really-satisfies?` them.

cgrand commented on src/unrepl/printer.clj:
```diff
@@ -34,18 +34,18 @@
-  (if-some [Datafiable (some-> 'clojure.core.protocols/Datafiable resolve deref)]
-    #(or (get (meta %) 'clojure.core.protocols/datafy) (really-satisfies? Datafiable %))
-    (constantly false)))
+  #(if-some [Datafiable (some-> 'clojure.core.protocols/Datafiable resolve deref)]
```
The problem is the `deref`, could you just remove it from the `if-some` and put it in the anon fn on the line below?

zampino replied:
ok

cgrand commented on src/unrepl/printer.clj:
```diff
@@ -34,18 +34,18 @@
       (clojure.lang.Var$Unbound. #'datafy)))
 
 (def ^:private navigable?
-  (if-some [Navigable (some-> 'clojure.core.protocols/Navigable resolve deref)]
```
same as above

zampino committed:
Check u.p/datafiable? against up-to-date implementations of the protocol

moving 'deref' at call time. Fixes datafiable? when protocol is extended 
after unrepl upgrade. Same for navigable?

zampino commented:
I moved the `deref` but kept the same variable names, maybe not so nice...

cgrand commented:
@zampino thanks
